### ISSUE TITLE
--jobs not respected in Bazel 7

### ISSUE THREAD

**Issue by chiragramani** [bazelbuild/bazel#20521]
### Description of the bug:

We have used `--jobs` to control the number of jobs that should be executed concurrently. This works as expected in Bazel 6 however, in Bazel 7, this functionality does not work as expected. 

### Which category does this issue belong to?

Core

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

1. This is a modified version of the official bazel examples.
[JobsRepro.zip](https://github.com/bazelbuild/bazel/files/13657549/JobsRepro.zip)
2. With bazel 7,
```
build :hello-time :hello-time1 :hello-time2 :hello-time3 :hello-time4
```
even though jobs=1 is specified in the .bazelrc, we still see parallel jobs > 1. 
![Screenshot 2023-12-13 at 12 57 12 PM](https://github.com/bazelbuild/bazel/assets/11925399/29d17aee-648a-4c1d-a4a2-d46a99208fc1)
3. But the same example, with Bazel 6.4, respects --jobs=1.
![Screenshot 2023-12-13 at 12 58 43 PM](https://github.com/bazelbuild/bazel/assets/11925399/96953f59-1ff7-4048-a183-78e3a0ff5d89)

4. With Bazel 7, if I use `--local_cpu_resources=1`, then it seems to work as expected. But from the documentation, --jobs is a global override to control such parallelism, wondering if this is a regression or if this behavior is driven differently in Bazel 7 intentionally. 

### Which operating system are you running Bazel on?

macOS

### What is the output of `bazel info release`?

7.0.0

**Comment by coeuvre** [comment]
It's due to the upgrade to JDK21 where the implementation of ForkJoinPool has changed -- the number of core threads is larger than `--jobs`. I can get the old behaviour by using `--experimental_use_semaphore_for_jobs` with your repro. We probably should flip and remove this flag.

**Comment by chiragramani** [comment]
> It's due to the upgrade to JDK21 where the implementation of ForkJoinPool has changed -- the number of core threads is larger than `--jobs`. I can get the old behaviour by using `--experimental_use_semaphore_for_jobs` with your repro. We probably should flip and remove this flag.

Thanks for looking into it. Yes, using `--experimental_use_semaphore_for_jobs` helped. Thanks for sharing this workaround!

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.1.0 RC1](https://releases.bazel.build/7.1.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

**Comment by hxt365** [comment]
Hi @chiragramani, can you share how you get the visualization? Thanks!
